### PR TITLE
Asyncify the API

### PR DESCRIPTION
**Pull request by bubonicbob**
Change the API to use asyncio to be compatible with Home Assistant.   This was discussed [here](https://github.com/home-assistant/core/issues/100065#issuecomment-1712918564).

### CODE REVIEW

**Comment by bubonicbob** [comment]
Please be kind, this is my first github contribution!  :)    Let me know how I can make this change better.

I don't have a powerwall myself so I haven't run the integration tests.   Is there anybody who would be willing to help me out and try it?

**Comment by ejbatts** [comment]
> Please be kind, this is my first github contribution! :) Let me know how I can make this change better.
> 
> I don't have a powerwall myself so I haven't run the integration tests. Is there anybody who would be willing to help me out and try it?

I can offer to assist with testing, but have no idea how. If you explain any procedure I will try.

**Comment by bubonicbob** [comment]
> > Please be kind, this is my first github contribution! :) Let me know how I can make this change better.
> > I don't have a powerwall myself so I haven't run the integration tests. Is there anybody who would be willing to help me out and try it?
> 
> I can offer to assist with testing, but have no idea how. If you explain any procedure I will try.

Thanks!   I'm on a mac, and I would run:
`git clone https://github.com/jrester/tesla_powerwall.git
cd tesla_powerwall
gh pr checkout 55
POWERWALL_IP=youraddress POWERWALL_PASSWORD=yourpass tox -e integration
`

If you're missing any commands like gh and tox, run:
`brew install gh tox`

**Comment by bubonicbob** [comment]
Hey @dh99999 would you be able to help test out this pull request for me?   Also could I get you to help me get refreshed fixture data.   Seems like there has been some big changes in the data structure as compared to version 0.3.19... maybe this corresponds to updated firmware.  I wonder why it isn't a breaking change since it deviates a lot from the fixture data in the powerwall HA integration.

**Comment by ejbatts** [comment]
> > > Please be kind, this is my first github contribution! :) Let me know how I can make this change better.
> > > I don't have a powerwall myself so I haven't run the integration tests. Is there anybody who would be willing to help me out and try it?
> > 
> > 
> > I can offer to assist with testing, but have no idea how. If you explain any procedure I will try.
> 
> Thanks! I'm on a mac, and I would run:
> 
> ```
> git clone https://github.com/jrester/tesla_powerwall.git
> cd tesla_powerwall
> gh pr checkout 55
> POWERWALL_IP=youraddress POWERWALL_PASSWORD=yourpass tox -e integration
> ```
> 
> If you're missing any commands like gh and tox, run:
> 
> ```
> brew install gh tox
> ```

**Comment by ejbatts** [comment]
I am on a PC, how do I install gh and tox? Brew is not recognised

**Comment by bubonicbob** [comment]
> I am on a PC, how do I install gh and tox? Brew is not recognised

You'll have to find a guide to get python and pip installed.   Then [install tox](https://tox.wiki/en/4.11.4/installation.html).

I've never tried getting gh for windows, but perhaps it is [available here](https://cli.github.com/)?   Else it might be easier to get this pull request via the [github desktop app](https://docs.github.com/en/desktop/working-with-your-remote-repository-on-github-or-github-enterprise/viewing-a-pull-request-in-github-desktop#opening-a-pull-request-branch-in-github-desktop-from-github).

**Comment by ejbatts** [comment]
got to the last command, tox has installed correctly but cant be found???

admin@STUDY MINGW64 ~/tesla_powerwall (asyncify)
$ powerwall_ip=192.168.1.41 powerwall_password=xxxxxx tox -e integration
****bash: tox: command not found****

File: tesla_powerwall/powerwall.py
Comment (jrester) [inline tesla_powerwall/powerwall.py:5]:
```suggestion
```
import not needed

File: pyproject.toml
Comment (jrester) [inline pyproject.toml:18]:
```suggestion
```
can be removed since we are using aiohttp now, right?

File: tests/unit/test_powerwall.py
Comment (jrester) [inline tests/unit/test_powerwall.py:5]:
```suggestion
import json
```

File: tests/unit/test_powerwall.py
Comment (jrester) [inline tests/unit/test_powerwall.py:58]:
```suggestion
    def add_response(self, path: str, method: str = "GET", body: Optional[dict] = None):
```
I would prefer to not use the name of the `json`, and instead use something else

**Comment by bubonicbob** [comment]
> got to the last command, tox has installed correctly but cant be found???
> 
> admin@STUDY MINGW64 ~/tesla_powerwall (asyncify) $ powerwall_ip=192.168.1.41 powerwall_password=xxxxxx tox -e integration **bash: tox: command not found**

Hmm, you can try this command and maybe the first line shows you the path to the tox binary?
```
pip show -f tox
```

**Review by jrester — CHANGES_REQUESTED** [review]
@bubonicbob Thank you very much for your PR. I think it already looks quite good, just some minor points that I would like to address before merging it.

**Comment by ejbatts** [comment]
> 
> Hmm, you can try this command and maybe the first line shows you the path to the tox binary?
> 
> ```
> pip show -f tox
> ```

Okay I can see the path but how do i incorporate that into the cmd  

`powerwall_ip=192.168.1.41 powerwall_password=xxxxxx tox -e integration `

**Comment by quozl** [comment]
Had a quick look, and saw nothing obviously wrong.  Here's what I thought of;

* if requests is no longer needed, remove from requirements?
* will merging this break all code that uses this module?  i.e. force asyncify of others.

**Comment by bubonicbob** [comment]
> > Hmm, you can try this command and maybe the first line shows you the path to the tox binary?
> > ```
> > pip show -f tox
> > ```
> 
> Okay I can see the path but how do i incorporate that into the cmd
> 
> `powerwall_ip=192.168.1.41 powerwall_password=xxxxxx tox -e integration `

`POWERWALL_IP` and `POWERWALL_PASSWORD` are just two individual environment variables (btw I bet these are case-sensitive so use all caps).   I gave you [a fancy way](https://stackoverflow.com/questions/1420719/powershell-setting-an-environment-variable-for-a-single-command-only) that bash can supply the variables temporarily at the same time, but you can follow guides for [bash](https://ioflood.com/blog/bash-set-environment-variable/#:~:text=Setting%20an%20environment%20variable%20in,export%20VARNAME%3D%22value%22%20.&text=In%20this%20example%2C%20we%20set,Anton%20using%20the%20export%20command), [powershell](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_environment_variables?view=powershell-7.4#using-the-variable-syntax), or [command prompt](https://learn.microsoft.com/en-us/windows-server/administration/windows-commands/set_1) to just set the variables by themselves so that they take effect when you run tox alone later in your shell.

It is a good habit to not set passwords into your environment so don't go out of your way to make these global settings, or if you really want to make sure, set your `POWERWALL_PASSWORD` to something bogus when you're done.

After setting those two variables for your specific shell, you can just run tox by itself:

```
tox -e integration
```

**Comment by bubonicbob** [comment]
> Had a quick look, and saw nothing obviously wrong. Here's what I thought of;
> 
> * if requests is no longer needed, remove from requirements?
> * will merging this break all code that uses this module?  i.e. force asyncify of others.

Thank you!   Yes merging this will break anything that depends on this library.   In my limited experience, it seems typical in python-land that libraries in this situation just hard fork into something with an 'a' in front of the original name and don't bother with maintaining a hybrid api.   In the case of this specific library, [there's support ](https://github.com/home-assistant/core/issues/100065#issuecomment-1712918564) for taking the plunge and moving the API over.

@jrester since this is a major API change, should the version change to 0.5.0?   Not sure the official way to do that.

**Comment by bubonicbob** [comment]
> @bubonicbob Thank you very much for your PR! I think it already looks quite good, just some minor points that I would like to address before merging it. Additionally, I will also do some testing of your implementation soon, to see if everything works as expected.
> 
> Nevertheless, I am quite unsure, about the use of [`aresponses` ](https://github.com/aresponses/aresponses) for testing, as the project seems to be inactive for over two years now and aiohttp looks like it [already has good support for testing](https://docs.aiohttp.org/en/stable/testing.html#testing). I would prefer the direct use of aiohttp, but I would see this more as a future improvement, than a requirement for this PR, as it might otherwise blow the scope. If you want to take a look, I would be happy to also merge this, otherwise I will take care of this in the future.

Much appreciated!  I'm pretty new to python so I don't know what are the popular choices for testing.   I picked [`aresponses` ](https://github.com/aresponses/aresponses) since the [tesla-wall-connector](https://github.com/einarhauks/tesla-wall-connector) similar library used it.   I was confused even more to learn that [HA wants to stop](https://github.com/home-assistant/core/issues/40855) using `unittest.TestCase` style tests.   It is a jungle out there.   🤷 

While you're testing, I'd really appreciate help updating the [fixture data for the HA integration tests](https://github.com/home-assistant/core/tree/dev/tests/components/powerwall/fixtures).   I'm in the middle of making a pull request to use this library and the fixture data breaks tests over there.   I saw [these samples](https://github.com/vloschiavo/powerwall2/tree/master/samples) from a similar project but still the powerwall firmware is a few years old.

**Comment by ejbatts** [comment]
I got tox to run by exporting the PATH using:
`export PATH=$PATH:/c/users/admin/AppData/Roaming/Python/Python312/scripts
` I note that tox.exe was located in 'scripts' and not 'site-packages' as suggested by previous analysis.
Having now successfully run the evaluation I have received the following summary:
_**integration: FAIL code 1 (11.91 seconds)
evaluation failed :( (12.05 seconds)**_
It appears that the error lies here:
**_integration: install_deps> python -I -m pip install aresponses
error: subprocess-exited-with-error_**

....so I am looking for more help please.

**Comment by bubonicbob** [comment]
> multidict

Weird.    The `multidict` package should have been installed automatically because there is a reference to aiohttp as a dependency.   I wonder if pip is having some issue fetching the dependencies?   Try running (from the root tesla_powerwall directory):

```
tox exec -- pip show aresponses
```

You can manually run the pip install for it like this:
```
tox exec -- pip install aresponses
tox exec -- pip install multidict
```

**Comment by ejbatts** [comment]
Tried both of the above but failed. Verbose output:
error: subprocess-exited-with-error

  Building wheel for multidict (pyproject.toml) did not run successfully.
  exit code: 1

  [74 lines of output]
  *********************
  * Accelerated build *
  *********************
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build\lib.win-amd64-cpython-312
  creating build\lib.win-amd64-cpython-312\multidict
  copying multidict\_abc.py -> build\lib.win-amd64-cpython-312\multidict
  copying multidict\_compat.py -> build\lib.win-amd64-cpython-312\multidict
  copying multidict\_multidict_base.py -> build\lib.win-amd64-cpython-312\multidict
  copying multidict\_multidict_py.py -> build\lib.win-amd64-cpython-312\multidict
  copying multidict\__init__.py -> build\lib.win-amd64-cpython-312\multidict
  running egg_info
  writing multidict.egg-info\PKG-INFO
  writing dependency_links to multidict.egg-info\dependency_links.txt
  writing top-level names to multidict.egg-info\top_level.txt
  reading manifest file 'multidict.egg-info\SOURCES.txt'
  reading manifest template 'MANIFEST.in'
  warning: no previously-included files matching '*.pyc' found anywhere in distribution
  warning: no previously-included files found matching 'multidict\_multidict.html'
  warning: no previously-included files found matching 'multidict\*.so'
  warning: no previously-included files found matching 'multidict\*.pyd'
  warning: no previously-included files found matching 'multidict\*.pyd'
  no previously-included directories found matching 'docs\_build'
  adding license file 'LICENSE'
  writing manifest file 'multidict.egg-info\SOURCES.txt'
  C:\Users\admin\AppData\Local\Temp\pip-build-env-v7vf87r2\overlay\Lib\site-packages\setuptools\command\build_py.py:207: _Warning: Package 'multidict._multilib' is absent from the `packages` configuration.
  !!

          ********************************************************************************
          ############################
          # Package would be ignored #
          ############################
          Python recognizes 'multidict._multilib' as an importable package[^1],
          but it is absent from setuptools' `packages` configuration.

          This leads to an ambiguous overall configuration. If you want to distribute this
          package, please make sure that 'multidict._multilib' is explicitly added
          to the `packages` configuration field.

**Comment by ejbatts** [comment]
Test still fails :(

**Comment by bubonicbob** [comment]
> POWERWALL_IP=youraddress POWERWALL_PASSWORD=yourpass tox -e integration

**Comment by ejbatts** [comment]
Attached output from last test:

[Verbose Output.txt](https://github.com/jrester/tesla_powerwall/files/13836426/Verbose.Output.txt)

**Comment by bubonicbob** [comment]
> Attached output from last test:
> 
> [Verbose Output.txt](https://github.com/jrester/tesla_powerwall/files/13836426/Verbose.Output.txt)

Thank you so much for collecting this!   Looks like maybe the tests are overwhelming your gateway, at least after the first test failed with a timeout.    I wonder if restarting the tests just once after waiting for some time would change the results or at least reveal the original bug that needs to be fixed.

**Review by jrester — APPROVED** [review]
Looks great! If the integration tests are also fine, this can be merged.

**Comment by ejbatts** [comment]
@bubonicbob ran test as per above and this seems to be a new error:

tesla_powerwall.error.AccessDeniedError: Access denied for resource https://192.168.1.41/api/system_status/soe: Unable to GET to resource: **User does not have adequate access rights**

[Verbose Output2.txt](https://github.com/jrester/tesla_powerwall/files/13845840/Verbose.Output2.txt)

**Comment by bubonicbob** [comment]
> @bubonicbob ran test as per above and this seems to be a new error:
> 
> tesla_powerwall.error.AccessDeniedError: Access denied for resource https://192.168.1.41/api/system_status/soe: Unable to GET to resource: **User does not have adequate access rights**
> 
> [Verbose Output2.txt](https://github.com/jrester/tesla_powerwall/files/13845840/Verbose.Output2.txt)
> 
> The comprehensive test using `POWERWALL_IP=youraddress POWERWALL_PASSWORD=yourpass tox -e integration` results in the same access error for all tests.
> 
> For clarification, i am running these tests within the Git Bash app for Windows, and I have two PW2 batteries connected to a single Gateway.

Ok let's try doing the same test without my changes.

First switch to another pull request that doesn't have these changes:
```
gh pr checkout 56
```

Then try running the same single test again.   Better not to run the full test suite until at least one passes.

**Comment by ejbatts** [comment]
@bubonicbob I think this is a PASS?
integration: recreate env because requirements removed: aresponses
integration: remove tox env folder C:\Users\admin\tesla_powerwall\.tox\integration
integration: install_deps> python -I -m pip install responses
.pkg: _optional_hooks> python C:\Users\admin\AppData\Roaming\Python\Python312\site-packages\pyproject_api\_backend.py True setuptools.build_meta
.pkg: get_requires_for_build_sdist> python C:\Users\admin\AppData\Roaming\Python\Python312\site-packages\pyproject_api\_backend.py True setuptools.build_meta
.pkg: build_sdist> python C:\Users\admin\AppData\Roaming\Python\Python312\site-packages\pyproject_api\_backend.py True setuptools.build_meta
integration: install_package_deps> python -I -m pip install requests>=2.22.0
integration: install_package> python -I -m pip install --force-reinstall --no-deps C:\Users\admin\tesla_powerwall\.tox\.tmp\package\17\tesla_powerwall-0.4.0.tar.gz
integration: commands[0]> python -m unittest tests/integration/test_powerwall.py -k test_get_charge
.
----------------------------------------------------------------------
Ran 1 test in 0.902s

OK
.pkg: _exit> python C:\Users\admin\AppData\Roaming\Python\Python312\site-packages\pyproject_api\_backend.py True setuptools.build_meta
  integration: OK (17.80=setup[14.80]+cmd[3.00] seconds)
  congratulations :) (17.95 seconds)

**Comment by ejbatts** [comment]
Just ran
gh pr checkout 56
POWERWALL_IP=youraddress POWERWALL_PASSWORD=yourpass tox -e integration
and got a PASS on 10 tests. The branch is (fix_integration), not (asyncify). 

Switched back to asyncify and received the access errors again.

**Comment by bubonicbob** [comment]
> Just ran gh pr checkout 56 POWERWALL_IP=youraddress POWERWALL_PASSWORD=yourpass tox -e integration and got a PASS on 10 tests. The branch is (fix_integration), not (asyncify).
> 
> Switched back to asyncify and received the access errors again.

I just updated the PR with a little extra to verify your authentication cookie in the integration tests.   If you pull it again you should get that update.   Please post updated test results with that change.   If it isn't authenticating and storing the results in a cookie, I'm not sure how to make progress.   I'll have to get help from another dev willing to troubleshoot it with a live powerwall.

**Comment by ejbatts** [comment]
Now returns an "AssertionError" for every test.

> Traceback (most recent call last):
>   File "C:\Users\admin\AppData\Local\Programs\Python\Python312\Lib\unittest\async_case.py", line 87, in _callSetUp
>     self._callAsync(self.asyncSetUp)
>   File "C:\Users\admin\AppData\Local\Programs\Python\Python312\Lib\unittest\async_case.py", line 104, in _callAsync
>     return self._asyncioRunner.run(
>            ^^^^^^^^^^^^^^^^^^^^^^^^
>   File "C:\Users\admin\AppData\Local\Programs\Python\Python312\Lib\asyncio\runners.py", line 118, in run
>     return self._loop.run_until_complete(task)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "C:\Users\admin\AppData\Local\Programs\Python\Python312\Lib\asyncio\base_events.py", line 684, in run_until_complete
>     return future.result()
>            ^^^^^^^^^^^^^^^
>   File "C:\Users\admin\tesla_powerwall\tests\integration\test_powerwall.py", line 20, in asyncSetUp
>     assert self.powerwall.is_authenticated()
> AssertionError
>

**Comment by bubonicbob** [comment]
> Now returns an "AssertionError" for every test.
> 
> > Traceback (most recent call last):
> > File "C:\Users\admin\AppData\Local\Programs\Python\Python312\Lib\unittest\async_case.py", line 87, in _callSetUp
> > self._callAsync(self.asyncSetUp)
> > File "C:\Users\admin\AppData\Local\Programs\Python\Python312\Lib\unittest\async_case.py", line 104, in _callAsync
> > return self._asyncioRunner.run(
> > ^^^^^^^^^^^^^^^^^^^^^^^^
> > File "C:\Users\admin\AppData\Local\Programs\Python\Python312\Lib\asyncio\runners.py", line 118, in run
> > return self._loop.run_until_complete(task)
> > ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> > File "C:\Users\admin\AppData\Local\Programs\Python\Python312\Lib\asyncio\base_events.py", line 684, in run_until_complete
> > return future.result()
> > ^^^^^^^^^^^^^^^
> > File "C:\Users\admin\tesla_powerwall\tests\integration\test_powerwall.py", line 20, in asyncSetUp
> > assert self.powerwall.is_authenticated()
> > AssertionError

@jrester Looks like I'm going to need help here.   For some reason the aiohttp session isn't storing the auth cookie from the successful login request.   @ejbatts testing showed that it works fine in the original `requests` based login but the same request in the `aiohttp` POST isn't storing the cookie.   I can't see anything in the documentation that suggests the default cookie jar behavior would be different between them and I don't see what would be different between how the code calls into the apis.

Not sure what else to try on my end without having a powerwall available to me.

**Comment by ejbatts** [comment]
> Looks like I'm going to need help here. For some reason the aiohttp session isn't storing the auth cookie from the successful login request. @ejbatts testing showed that it works fine in the original requests based login but the same request in the aiohttp POST isn't storing the cookie.


Just to be sure @bubonicbob , It appears to not be a Access Error anymore, but an Assertion Error. I have attached the output from the single test. 

[Verbose Output3.txt](https://github.com/jrester/tesla_powerwall/files/13847139/Verbose.Output3.txt)

**Comment by bubonicbob** [comment]
> > Looks like I'm going to need help here. For some reason the aiohttp session isn't storing the auth cookie from the successful login request. @ejbatts testing showed that it works fine in the original requests based login but the same request in the aiohttp POST isn't storing the cookie.
> 
> Just to be sure @bubonicbob , It appears to not be a Access Error anymore, but an Assertion Error. I have attached the output from the single test.
> 
> [Verbose Output3.txt](https://github.com/jrester/tesla_powerwall/files/13847139/Verbose.Output3.txt)

Yes this new error makes sense since I added a new check to see if the auth cookie had been stored successfully.  Seems to not have been stored which would have caused the unauthorized error you had seen before I made this change.  Without the cookie it would have been like you had never logged in.

Your test against master proved that your gateway is sending this cookie.  I'm at a loss as to what's going on.

**Comment by bubonicbob** [comment]
> > Looks like I'm going to need help here. For some reason the aiohttp session isn't storing the auth cookie from the successful login request. @ejbatts testing showed that it works fine in the original requests based login but the same request in the aiohttp POST isn't storing the cookie.
> 
> Just to be sure @bubonicbob , It appears to not be a Access Error anymore, but an Assertion Error. I have attached the output from the single test.
> 
> [Verbose Output3.txt](https://github.com/jrester/tesla_powerwall/files/13847139/Verbose.Output3.txt)

I think I figured it out.   The issue is that, by default, aiohttp ignores cookies when your POWERWALL_IP is an IP address instead of a DNS name.   If you pull the latest updates, I think it will start working.

**Comment by ejbatts** [comment]
Just ran 
`gh pr checkout 55`
and the individual test...same result...
 

> File "C:\Users\admin\tesla_powerwall\tests\integration\test_powerwall.py", line 20, in asyncSetUp
>     assert self.powerwall.is_authenticated()
> AssertionError
>

**Comment by bubonicbob** [comment]
> Just ran `gh pr checkout 55` and the individual test...same result...
> 
> > File "C:\Users\admin\tesla_powerwall\tests\integration\test_powerwall.py", line 20, in asyncSetUp
> > assert self.powerwall.is_authenticated()
> > AssertionError

My fault, please try one more time.

**Comment by ejbatts** [comment]
> My fault, please try one more time.

SUCCESS!

Ran 10 tests in 27.502s

OK
.pkg: _exit> python C:\Users\admin\AppData\Roaming\Python\Python312\site-packages\pyproject_api\_backend.py True setuptools.build_meta
  integration: OK (36.28=setup[8.34]+cmd[27.94] seconds)
  congratulations :) (36.45 seconds)

**Comment by bubonicbob** [comment]
> [@ejbatts](/ejbatts)
> ###

@ejbatts Sweet.   I owe you a beer next time you're in Seattle.   🍻

**Comment by ejbatts** [comment]
Will the changed integration allow me to access individual battery data? The essential thing I want is   BATTERY POWER (charge/discharge rate) as individual entities and not the combined value that Tesla provides as default.

**Comment by bubonicbob** [comment]
> Will the changed integration allow me to access individual battery data? The essential thing I want is BATTERY POWER (charge/discharge rate) as individual entities and not the combined value that Tesla provides as default.

Create an issue on this repo for that and we can all discuss there.   [Instructions](https://docs.github.com/en/issues/tracking-your-work-with-issues/creating-an-issue).

**Comment by ejbatts** [comment]
Oh okay. I actually thought this testing was heading in that direction after I posted  this on [https://github.com/home-assistant/core/issues/100065#top](url) thread.

> I would like a similar feature but the opposite way round. Is it possible to extract data from separate powerwalls attached to a single gateway? I have 2 powerwalls connected to separate phases of a 3 phase installation. Specifically, I would like to know battery load (charge/discharge rate) of each battery so I can monitor house loads on each phase.
> This separate data is available on //[gateway ip address]/system page, as well as voltage and state of charge, but I don't know how to extract it in HA.

I'll raise a blank issue.

File: tesla_powerwall/api.py
Comment (bdraco) [inline tesla_powerwall/api.py:176]:
```suggestion
    async def close(self) -> None:
```

mypy complained this was untyped

File: tesla_powerwall/powerwall.py
Comment (bdraco) [inline tesla_powerwall/powerwall.py:203]:
```suggestion
    async def close(self) -> None:
```

mypy complained this was untyped

File: tesla_powerwall/powerwall.py
Comment (bdraco) [inline tesla_powerwall/powerwall.py:206]:
```suggestion
    async def __aenter__(self) -> "Powerwall":
```

File: tesla_powerwall/api.py
Comment (bdraco) [inline tesla_powerwall/api.py:29]: I think we need the unsafe cookie jar to allow the ip based cookies here as well
https://github.com/home-assistant/core/pull/107164/files#diff-f1eecf25dc296b0bce35cd4f3f74c6b88f3396694f986c97d804bf4a33a62b97R125

**Comment by bdraco** [comment]
Everything seems to be working great after I small fixes I added to https://github.com/home-assistant/core/pull/107164

**Comment by bdraco** [comment]
Pretty significant drop in cpu usage for the HA integration observed as well

**Comment by bdraco** [comment]
There's still a lot of JSON being passed around, which is what's mostly left of the CPU runtime, using orjson to parse would probably solve that, but that's for another PR.

**Comment by bdraco** [comment]
https://github.com/bdraco/nexia/blob/ee3253ab8cf1f128b700e59bb09354e21612851e/nexia/home.py#L244

File: tesla_powerwall/powerwall.py
Comment (bubonicbob) [reply tesla_powerwall/powerwall.py:203]: Does that mean it exported the typing info for you?   I'm still wondering why it didn't do it for me and I had to add all the cast calls in the HA core.

**Comment by bubonicbob** [comment]
> There's still a lot of JSON being passed around, which is what's mostly left of the CPU runtime, using orjson to parse would probably solve that, but that's for another PR.

Added an [issue](https://github.com/jrester/tesla_powerwall/issues/58) for it.

File: tesla_powerwall/powerwall.py
Comment (bdraco) [reply tesla_powerwall/powerwall.py:203]: I would assume its seeing the typing based on it warning

**Comment by bdraco** [comment]
I think this looks ready for a release once the typing comments and the aiohttp session comments are sorted

The HA PR looks like its good to go after this is released and the tests are updated

**Comment by bubonicbob** [comment]
> I think this looks ready for a release once the typing comments and the aiohttp session comments are sorted
> 
> The HA PR looks like its good to go after this is released and the tests are updated

I'm ready to get this released.   I don't think I'm able to do this myself.

File: tox.ini
Comment (jrester) [inline tox.ini:28]: This is already handled by pre-commit. Please remove this from tox and use pre-commit instead.

**Review by jrester — CHANGES_REQUESTED** [review]

**Comment by bdraco** [comment]
Retested with latest code. Everything seems to be working as expected

**Review by jrester — APPROVED** [review]

**Comment by jrester** [comment]
Thank you very much @bubonicbob for implementing this. Also thank you very much @bdraco, @ejbatts and @quozl for testing and reviewing!

**Comment by jrester** [comment]
Version 0.5.0 is now live on PyPi

**Comment by quozl** [comment]
No worries.  Well done everyone.

**Pull request merged**
